// notion-compat/build/index.js
// src/notion-compat-api.ts
import { parsePageId } from "notion-utils";
import PQueue from "p-queue";

// src/convert-color.ts
function convertColor(color) {
  switch (color) {
    case "green":
      return "teal";
    case "green_background":
      return "teal_background";
    default:
      return color;
  }
}

// src/convert-rich-text.ts
function convertRichText(richText) {
  return richText.map(convertRichTextItem).filter(Boolean);
}
function convertRichTextItem(richTextItem) {
  var _a;
  const subdecorations = [];
  if (richTextItem.annotations.bold) {
    subdecorations.push(["b"]);
  }
  if (richTextItem.annotations.italic) {
    subdecorations.push(["i"]);
  }
  if (richTextItem.annotations.strikethrough) {
    subdecorations.push(["s"]);
  }
  if (richTextItem.annotations.underline) {
    subdecorations.push(["_"]);
  }
  if (richTextItem.annotations.code) {
    subdecorations.push(["c"]);
  }
  if (richTextItem.annotations.color !== "default") {
    subdecorations.push(["h", convertColor(richTextItem.annotations.color)]);
  }
  const details = richTextItem[richTextItem.type];
  if (details) {
    if (details.link) {
      subdecorations.push(["a", details.link.url]);
    }
  }
  switch (richTextItem.type) {
    case "text": {
      if (subdecorations.length) {
        return [richTextItem.text.content, subdecorations];
      } else {
        return [richTextItem.text.content];
      }
    }
    case "equation":
      if ((_a = richTextItem.equation) == null ? void 0 : _a.expression) {
        subdecorations.unshift(["e", richTextItem.equation.expression]);
      }
      return ["\u204D", subdecorations];
    case "mention": {
      const { mention } = richTextItem;
      if (mention) {
        switch (mention.type) {
          case "link_preview":
            subdecorations.push(["a", mention.link_preview.url]);
            break;
          case "page":
            subdecorations.push(["p", mention.page.id]);
            return ["\u2023", subdecorations];
          case "database":
            subdecorations.push(["p", mention.database.id]);
            return ["\u2023", subdecorations];
          case "date":
            subdecorations.unshift([
              "d",
              {
                type: "date",
                start_date: mention.date.start,
                end_date: mention.date.end,
                time_zone: mention.date.time_zone
              }
            ]);
            break;
          case "user":
            subdecorations.push(["u", mention.user.id]);
            break;
          case "template_mention":
            break;
          default:
            break;
        }
      }
      return [richTextItem.plain_text, subdecorations];
    }
    default:
      return [""];
  }
}

// src/convert-time.ts
function convertTime(time) {
  if (time) {
    try {
      return new Date(time).getTime();
    } catch {
    }
  }
  return void 0;
}

// src/convert-block.ts
function convertBlock({
  block: partialBlock,
  children = [],
  pageMap,
  blockMap,
  parentMap
}) {
  var _a, _b, _c, _d, _e, _f, _g, _h, _i, _j, _k, _l;
  const compatBlock = {
    id: partialBlock.id
  };
  if (children && children.length) {
    compatBlock.content = children;
  }
  const block = partialBlock;
  if (!block.type) {
    return compatBlock;
  }
  compatBlock.properties = {};
  compatBlock.format = {};
  compatBlock.type = block.type;
  compatBlock.created_time = convertTime(block.created_time);
  compatBlock.last_edited_time = convertTime(block.last_edited_time);
  if (block.created_by) {
    compatBlock.created_by_table = `notion_${block.created_by.object}`;
    compatBlock.created_by_id = block.created_by.id;
  }
  compatBlock.last_edited_by_table = (_a = block.last_edited_by) == null ? void 0 : _a.object;
  compatBlock.last_edited_by_id = (_b = block.last_edited_by) == null ? void 0 : _b.id;
  compatBlock.alive = block.archived !== true;
  if (parentMap) {
    const parentId = parentMap[block.id];
    if (parentId) {
      compatBlock.parent_id = parentId;
      const parentBlock = blockMap == null ? void 0 : blockMap[parentId];
      if (parentBlock) {
        switch (parentBlock.type) {
          case "child_database":
            compatBlock.parent_table = "table";
            break;
          case "child_page":
          default:
            compatBlock.parent_table = "block";
            break;
        }
      } else {
        const parentPage = pageMap == null ? void 0 : pageMap[parentId];
        if (parentPage) {
          compatBlock.parent_table = "block";
        }
      }
    }
  }
  const blockDetails = block[block.type];
  if (blockDetails) {
    if (blockDetails.rich_text) {
      compatBlock.properties.title = convertRichText(blockDetails.rich_text);
    }
    if (blockDetails.color) {
      compatBlock.format.block_color = convertColor(blockDetails.color);
    }
    if (blockDetails.icon) {
      switch (blockDetails.icon.type) {
        case "emoji":
          compatBlock.format.page_icon = blockDetails.icon.emoji;
          break;
        case "external":
          compatBlock.format.page_icon = blockDetails.icon.external.url;
          break;
        case "file":
          compatBlock.format.page_icon = blockDetails.icon.file.url;
          break;
      }
    }
    if (blockDetails.type) {
      switch (blockDetails.type) {
        case "external":
          compatBlock.properties.source = [[blockDetails.external.url]];
          break;
        case "file":
          compatBlock.properties.source = [[blockDetails.file.url]];
          break;
      }
    }
  }
  switch (block.type) {
    case "paragraph":
      compatBlock.type = "text";
      if (!((_d = (_c = block.paragraph) == null ? void 0 : _c.rich_text) == null ? void 0 : _d.length)) {
        delete compatBlock.properties;
      }
      break;
    case "heading_1":
      compatBlock.type = "header";
      break;
    case "heading_2":
      compatBlock.type = "sub_header";
      break;
    case "heading_3":
      compatBlock.type = "sub_sub_header";
      break;
    case "bulleted_list_item":
      compatBlock.type = "bulleted_list";
      break;
    case "numbered_list_item":
      compatBlock.type = "numbered_list";
      break;
    case "quote":
      break;
    case "to_do":
      if ((_e = block.to_do) == null ? void 0 : _e.checked) {
        compatBlock.properties.checked = [["Yes"]];
      }
      break;
    case "toggle":
      break;
    case "code":
      if (block.code.language) {
        compatBlock.properties.language = [[block.code.language]];
      }
      break;
    case "callout":
      break;
    case "file":
      break;
    case "divider":
      break;
    case "breadcrumb":
      break;
    case "table_of_contents":
      break;
    case "column_list":
      break;
    case "column":
      break;
    case "bookmark":
      if (block.bookmark.url) {
        compatBlock.properties.link = [[block.bookmark.url]];
      }
      if (block.bookmark.caption) {
        compatBlock.properties.description = convertRichText(
          block.bookmark.caption
        );
      }
      break;
    case "link_to_page":
      compatBlock.type = "alias";
      switch ((_f = block.link_to_page) == null ? void 0 : _f.type) {
        case "page_id":
          compatBlock.format.alias_pointer = {
            id: block.link_to_page.page_id,
            table: "block"
          };
          break;
        case "database_id":
          compatBlock.format.alias_pointer = {
            id: block.link_to_page.database_id,
            table: "table"
          };
          break;
      }
      break;
    case "child_page": {
      compatBlock.type = "page";
      if (pageMap) {
        const page = pageMap[block.id];
        if (page) {
          if (page.properties.title) {
            compatBlock.properties.title = convertRichText(
              page.properties.title.title
            );
          }
          if (page.cover) {
            switch (page.cover.type) {
              case "external":
                compatBlock.format.page_cover = page.cover.external.url;
                break;
              case "file":
                compatBlock.format.page_cover = page.cover.file.url;
                break;
            }
            compatBlock.format.page_cover_position = 0.5;
          }
          if (page.icon) {
            switch (page.icon.type) {
              case "emoji":
                compatBlock.format.page_icon = page.icon.emoji;
                break;
              case "external":
                compatBlock.format.page_icon = page.icon.external.url;
                break;
              case "file":
                compatBlock.format.page_icon = page.icon.file.url;
                break;
            }
          }
          if (page.parent) {
            switch (page.parent.type) {
              case "workspace":
                compatBlock.parent_table = "space";
                break;
              case "database_id":
                compatBlock.parent_table = "table";
                break;
              case "page_id":
                compatBlock.parent_table = "block";
                break;
            }
          }
        }
      }
      if (block.child_page) {
        if (block.child_page.title) {
          compatBlock.properties.title = [[block.child_page.title]];
        }
      }
      break;
    }
    case "template":
      break;
    case "synced_block":
      if (block.synced_block.synced_from) {
        compatBlock.type = "transclusion_reference";
        compatBlock.format.transclusion_reference_pointer = {
          id: block.synced_block.synced_from.block_id,
          table: "block"
        };
      } else {
        compatBlock.type = "transclusion_container";
      }
      break;
    case "equation":
      if ((_g = block.equation) == null ? void 0 : _g.expression) {
        compatBlock.properties.title = [[block.equation.expression]];
      }
      break;
    case "child_database":
      break;
    case "table":
      if (blockDetails.table_width && blockDetails.table_width > 0) {
        compatBlock.format.table_block_column_order = Array.from(
          { length: blockDetails.table_width },
          (_, i) => "" + i
        );
        compatBlock.format.table_block_column_format = compatBlock.format.table_block_column_order.map((order) => {
          return {
            [order]: {}
          };
        });
      }
      if (blockDetails.has_column_header) {
        compatBlock.format.table_block_column_header = blockDetails.has_column_header;
      }
      if (blockDetails.has_row_header) {
        compatBlock.format.table_block_row_header = blockDetails.has_row_header;
      }
      break;
    case "table_row":
      compatBlock.properties = {
        ...(_i = (_h = block.table_row) == null ? void 0 : _h.cells) == null ? void 0 : _i.map((cell) => convertRichText(cell))
      };
      break;
    case "pdf":
      compatBlock.format.block_page_width = true;
      compatBlock.format.block_height = "80vh";
      break;
    case "video": {
      compatBlock.format.block_page_width = true;
      compatBlock.format.block_aspect_ratio = 0.5620608899297423;
      try {
        const url = (_k = (_j = compatBlock.properties.source) == null ? void 0 : _j[0]) == null ? void 0 : _k[0];
        if (!url)
          break;
        const u = new URL(url);
        switch (u.hostname) {
          case "loom.com":
          case "www.loom.com":
            u.pathname = u.pathname.replace(/^\/share\//i, "/embed/");
            compatBlock.format.display_source = u.toString();
            break;
        }
      } catch {
      }
      break;
    }
    case "embed": {
      const url = (_l = block.embed) == null ? void 0 : _l.url;
      compatBlock.format.block_page_width = true;
      compatBlock.format.block_height = "30vh";
      if (url) {
        compatBlock.properties.source = [[url]];
        try {
          const u = new URL(url);
          switch (u.hostname) {
            case "twitter.com":
              compatBlock.type = "tweet";
              delete compatBlock.format;
              break;
            case "maps.google.com":
              compatBlock.type = "maps";
              break;
            case "excalidraw.com":
              compatBlock.type = "excalidraw";
              break;
            case "codepen.io":
              compatBlock.type = "codepen";
              break;
            case "docs.google.com":
            case "drive.google.com":
              compatBlock.type = "drive";
              break;
            case "figma.com":
              compatBlock.type = "figma";
              break;
            case "open.spotify.com":
              if (u.pathname.includes("/embed/") || u.pathname.includes("/embed-podcast/")) {
                break;
              }
              if (u.pathname.startsWith("/playlist/")) {
                u.pathname = `/embed${u.pathname}`;
              } else if (u.pathname.startsWith("/episode/")) {
                u.pathname = `/embed-podcast${u.pathname}`;
              }
              u.search = "";
              compatBlock.format.display_source = u.toString();
              break;
            case "airtable.com":
              if (!u.pathname.startsWith("/embed/")) {
                u.pathname = `/embed${u.pathname}`;
                compatBlock.format.display_source = u.toString();
              }
              break;
            case "soundcloud.com":
              break;
          }
        } catch {
        }
      }
      break;
    }
    case "image":
      break;
    case "audio":
      break;
    case "link_preview":
      break;
    case "unsupported":
      break;
  }
  return compatBlock;
}

// src/convert-page.ts
function convertPage({
  pageId,
  blockMap,
  blockChildrenMap,
  pageMap,
  parentMap
}) {
  const compatBlocks = Object.values(blockMap).map(
    (block) => convertBlock({
      block,
      children: blockChildrenMap[block.id],
      pageMap,
      blockMap,
      parentMap
    })
  );
  const compatPageBlock = convertPageBlock({
    pageId,
    blockMap,
    blockChildrenMap,
    pageMap,
    parentMap
  });
  const compatPageBlocks = Object.keys(pageMap).filter((id) => id !== pageId).map(
    (id) => convertPageBlock({
      pageId: id,
      blockMap,
      blockChildrenMap,
      pageMap,
      parentMap
    })
  );
  const compatBlockMap = [
    compatPageBlock,
    ...compatBlocks,
    ...compatPageBlocks
  ].reduce(
    (blockMap2, block) => ({
      ...blockMap2,
      [block.id]: {
        type: "reader",
        value: block
      }
    }),
    {}
  );
  return {
    block: compatBlockMap,
    collection: {},
    collection_view: {},
    collection_query: {},
    signed_urls: {},
    notion_user: {}
  };
}
function convertPageBlock({
  pageId,
  blockMap,
  blockChildrenMap,
  pageMap,
  parentMap
}) {
  const partialPage = pageMap[pageId];
  const page = partialPage;
  if (page) {
    const compatPageBlock = convertBlock({
      block: { ...page, type: "child_page" },
      children: blockChildrenMap[page.id],
      pageMap,
      blockMap,
      parentMap
    });
    return compatPageBlock;
  }
  return null;
}

// src/notion-compat-api.ts
var NotionCompatAPI = class {
  constructor(client) {
    this.client = client;
  }
  async getPage(rawPageId) {
    const pageId = parsePageId(rawPageId);
    const [page, block, children] = await Promise.all([
      this.client.pages.retrieve({ page_id: pageId }),
      this.client.blocks.retrieve({ block_id: pageId }),
      this.getAllBlockChildren(pageId)
    ]);
    const { blockMap, blockChildrenMap, pageMap, parentMap } = await this.resolvePage(pageId);
    const recordMap = convertPage({
      pageId,
      blockMap,
      blockChildrenMap,
      pageMap,
      parentMap
    });
    recordMap.raw = {
      page,
      block,
      children
    };
    return recordMap;
  }
  async resolvePage(rootBlockId, {
    concurrency = 4
  } = {}) {
    const blockMap = {};
    const pageMap = {};
    const parentMap = {};
    const blockChildrenMap = {};
    const pendingBlockIds = /* @__PURE__ */ new Set();
    const queue = new PQueue({ concurrency });
    const processBlock = async (blockId, { shallow = false } = {}) => {
      if (!blockId || pendingBlockIds.has(blockId)) {
        return;
      }
      pendingBlockIds.add(blockId);
      queue.add(async () => {
        var _a, _b, _c;
        try {
          let partialBlock = blockMap[blockId];
          if (!partialBlock) {
            partialBlock = await this.client.blocks.retrieve({
              block_id: blockId
            });
            blockMap[blockId] = partialBlock;
          }
          const block = partialBlock;
          if (block.type === "child_page") {
            if (!pageMap[blockId]) {
              const partialPage = await this.client.pages.retrieve({
                page_id: blockId
              });
              pageMap[blockId] = partialPage;
              const page = partialPage;
              switch ((_a = page.parent) == null ? void 0 : _a.type) {
                case "page_id":
                  processBlock(page.parent.page_id, {
                    shallow: true
                  });
                  if (!parentMap[blockId]) {
                    parentMap[blockId] = page.parent.page_id;
                  }
                  break;
                case "database_id":
                  processBlock(page.parent.database_id, {
                    shallow: true
                  });
                  if (!parentMap[blockId]) {
                    parentMap[blockId] = page.parent.database_id;
                  }
                  break;
              }
            }
            if (blockId !== rootBlockId) {
              return;
            }
          }
          if (shallow) {
            return;
          }
          const children = await this.getAllBlockChildren(blockId);
          blockChildrenMap[blockId] = children.map((child) => child.id);
          for (const child of children) {
            const childBlock = child;
            const mappedChildBlock = blockMap[child.id];
            if (!mappedChildBlock || !mappedChildBlock.type && childBlock.type) {
              blockMap[child.id] = childBlock;
              parentMap[child.id] = blockId;
              const details = childBlock[childBlock.type];
              if (details == null ? void 0 : details.rich_text) {
                const richTextMentions = details.rich_text.filter(
                  (richTextItem) => richTextItem.type === "mention"
                );
                for (const richTextMention of richTextMentions) {
                  switch ((_b = richTextMention.mention) == null ? void 0 : _b.type) {
                    case "page": {
                      const pageId = richTextMention.mention.page.id;
                      processBlock(pageId, { shallow: true });
                      break;
                    }
                    case "database": {
                      const databaseId = richTextMention.mention.database.id;
                      processBlock(databaseId, { shallow: true });
                      break;
                    }
                  }
                }
              }
              if (childBlock.type === "link_to_page") {
                switch ((_c = childBlock.link_to_page) == null ? void 0 : _c.type) {
                  case "page_id":
                    processBlock(childBlock.link_to_page.page_id, {
                      shallow: true
                    });
                    break;
                  case "database_id":
                    processBlock(childBlock.link_to_page.database_id, {
                      shallow: true
                    });
                    break;
                }
              }
              if (childBlock.has_children && childBlock.type !== "child_database") {
                processBlock(childBlock.id);
              }
            }
          }
        } catch (err) {
          console.warn("failed resolving block", blockId, err.message);
        } finally {
          pendingBlockIds.delete(blockId);
        }
      });
    };
    await processBlock(rootBlockId);
    await queue.onIdle();
    return {
      blockMap,
      blockChildrenMap,
      pageMap,
      parentMap
    };
  }
  async getAllBlockChildren(blockId) {
    let blocks = [];
    let cursor;
    do {
      console.log("blocks.children.list", { blockId, cursor });
      const res = await this.client.blocks.children.list({
        block_id: blockId,
        start_cursor: cursor
      });
      blocks = blocks.concat(res.results);
      cursor = res.next_cursor;
    } while (cursor);
    return blocks;
  }
};
export {
  NotionCompatAPI,
  convertBlock,
  convertColor,
  convertPage,
  convertPageBlock,
  convertRichText,
  convertRichTextItem,
  convertTime
};
